storagemarket/provider_loop.go

```diff
@@ -97,16 +84,25 @@ func (p *Provider) loop() {
 			p.wg.Add(1)
 			go func() {
 				defer p.wg.Done()
-
-				p.doDeal(deal)
+				p.doDeal(deal, dealReq.dh)
 			}()
 
-			go writeDealResp(true, nil, nil)
+			dealReq.rsp <- acceptDealResp{true, nil, nil}
 			log.Infow("deal execution started", "id", deal.DealUuid)
 
-		case failedDeal := <-p.failedDealsChan:
-			log.Errorw("deal failed", "id", failedDeal.st.DealUuid, "err", failedDeal.err)
-			// Release storage space , funds, shared resources etc etc.
+		case finishedDeal := <-p.finishedDealsChan:
+			deal := finishedDeal.deal
+			log.Errorw("deal finished", "id", deal.DealUuid)
+			errf := p.fundManager.UntagFunds(p.ctx, deal.DealUuid)
+			if errf != nil {
+				log.Errorw("untagging funds", "id", deal.DealUuid, "err", errf)
+			}
+
+			errs := p.storageManager.Untag(p.ctx, deal.DealUuid)
+			if errs != nil {
+				log.Errorw("untagging storage", "id", deal.DealUuid, "err", errs)
+			}
```

Comment:
Can this code be extracted to a common function?

Reply:
@dirkmc I am thinking of refactoring the storage & fund manager to a "resource manager". Will discuss this offline and do this change in a follow up PR as part of it.

Reply:
Ok sounds good